modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -65,3 +73,63 @@ object GlobalSnapshotStorage {
   case class InvalidGlobalSnapshotChain(expected: GlobalSnapshotLink, actual: GlobalSnapshotLink) extends NoStackTrace
 
 }
+
+object GlobalSnapshotStorageV2 {
+
+  def make[F[_]: Async: KryoSerializer](
+    genesis: GlobalSnapshot,
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorageV2[F]] = {
+    def mkCacheQueue = Queue.unbounded[F, GlobalSnapshot]
+    def mkHeadRef = Ref.of[F, GlobalSnapshot](genesis)
+    def mkCache = MapRef.ofSingleImmutableMap[F, SnapshotOrdinal, GlobalSnapshot](Map.empty)
+    def mkOffloadQueue = Queue.unbounded[F, SnapshotOrdinal]
+
+    (mkCacheQueue, mkHeadRef, mkCache, mkOffloadQueue).mapN { make(_, _, _, _, inMemoryCapacity) }.flatten
+  }
+
+  def make[F[_]: Async: KryoSerializer](
+    cacheQueue: Queue[F, GlobalSnapshot],
+    headRef: Ref[F, GlobalSnapshot],
+    cache: MapRef[F, SnapshotOrdinal, Option[GlobalSnapshot]],
+    offloadQueue: Queue[F, SnapshotOrdinal],
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorageV2[F]] = {
+
+    def cacheProcess =
+      Stream
+        .fromQueueUnterminated(cacheQueue)
+        .evalTap { snapshot =>
+          headRef.set(snapshot) >>
+            cache(snapshot.ordinal).set(snapshot.some) >>
+            NonNegLong
+              .from(snapshot.ordinal.value.value - inMemoryCapacity.value)
```

Comment:
I think that if we calculate the next `SnapshotOrdinal` using `Next[A]`, we should also calculate the previous values with `Previous[A]` or `PartialPrevious[A]` (the latter one makes more sense for types like `NonNegLong`). Maybe also a syntax `previousN`/`partialPreviousN` could be added, analogous to [nextN](https://github.com/Constellation-Labs/tessellation/blob/c9d24567f4f7ac24e4b85b73586065384b2e4d4d/modules/shared/src/main/scala/org/tessellation/ext/cats/syntax/NextSyntax.scala#L20).